src/ParticleTracer.tpp

```diff
@@ -0,0 +1,104 @@
+
+        auto particle_orig = ptcls_->template get<0>();
+        auto particle_dest = ptcls_->template get<1>();
+        auto particle_ids = ptcls_->template get<2>();
+
+        bool success = pumipic::trace_particle_through_mesh(oh_mesh_, ptcls_, particle_orig, particle_dest,
+                                                            particle_ids, elem_ids_, next_elem_ids_,
+                                                            true, inter_faces_, inter_points_, last_exits_, 1000, true,
+                                                            func_, elmAreas_, ptcl_done_,
+                                                            tolerance_);
+
+        if (!success) {
+            printf("[ERROR] ParticleTracer: Failed to trace particles through the mesh.\n");
+            return false;
+        }
+        return true;
+    }
+
+
+    Omega_h::Write<Omega_h::LO> GetElemIds() { return elem_ids_; }
+
+    Omega_h::Write<Omega_h::LO> GetInterFaces() { return inter_faces_; }
+
+    Omega_h::Write<Omega_h::Real> GetInterPoints() { return inter_points_; }
+
+    Omega_h::Write<Omega_h::LO> GetLastExits() { return last_exits_; }
+
+private:
+    double tolerance_ = 1e-10;
+
+    Omega_h::Mesh &oh_mesh_;
+    pumipic::ParticleStructure<ParticleType> *ptcls_;
+    Func &func_;
+
+    Omega_h::Write<Omega_h::LO> elem_ids_;
+    Omega_h::Write<Omega_h::LO> next_elem_ids_;
+    Omega_h::Write<Omega_h::LO> inter_faces_;
+    Omega_h::Write<Omega_h::Real> inter_points_;
+    Omega_h::Write<Omega_h::LO> last_exits_;
+    Omega_h::Reals elmAreas_;
+    Omega_h::Write<Omega_h::LO> ptcl_done_;
+
+
+    bool validate_internal_data_sizes() {
+        int dim = oh_mesh_.dim();
+
+        if (elem_ids_.size() != next_elem_ids_.size() || elem_ids_.size() != inter_faces_.size() ||
+            elem_ids_.size() != inter_points_.size() / dim || elem_ids_.size() != last_exits_.size() ||
+            elem_ids_.size() != ptcl_done_.size()) {
+            printf("[ERROR] ParticleTracer: internal data arrays are not of the appropriate size.\nSize of elem_ids: %d, "
+                   "next_elem_ids: %d, inter_faces: %d, inter_points: %d last_exits: %d ptcl_done: %d\n",
+                   elem_ids_.size(), next_elem_ids_.size(), inter_faces_.size(), inter_points_.size(),
+                   last_exits_.size(),
+                   ptcl_done_.size());
+            return false;
+        }
+
+        // has to be either 0 or equal to the capacity of the particle structure
+        if ((elem_ids_.size() != 0) && (elem_ids_.size() != ptcls_->capacity())) {
+            printf("[ERROR] ParticleTracer: internal data arrays are not of appropriate size.\nSize of elem_ids has to"
+                   " be equal to the capacity of the particle structure.\nSize of elem_ids: %d, capacity: %d\n",
+                   elem_ids_.size(), ptcls_->capacity());
+            return false;
+        }
+
+        return true;
+    }
+
+    void reset_particle_done() {
```

Comment:
The name `done` is still a bit confusing to me. What about `particle_requires_tracing` or `particle_at_final_location`?